### PR TITLE
Port #49063 to master

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
Port #49063 to master

### What issues does this PR fix or reference?
Fixes: N/A

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/cloud/clouds/azurearm.py [hunk 1]
```diff
@@ -260,6 +263,14 @@ def __is_provider_configured(opts, provider, required_keys=()):
...
         )
 
+    if provider is False:
+        # check if using MSI style credentials...
+        provider = config.is_provider_configured(
+            __opts__,
+            __active_provider_name__ or __virtualname__,
+            required_keys=("subscription_id",),
+        )
+
```
Comment (waynew) [inline salt/cloud/clouds/azurearm.py:273]: From your question in Slack - we should probably refactor this code. But at minimum, we need to test:

1. if provider is found by `("subscription_id", "tenant", "client_id", "secret")`, that provider is returned
2. if provider is *not* found that way, but is found via `("subscription_id", "username", "password")`, it is returned
3. if provider is not found either of those ways, then whatever is returned from `config.is_provider_configured(...)` is returned.

Presumably that's False, but it would be more consistent and expected for these things to return None than False.

File: salt/cloud/clouds/azurearm.py [hunk 2]
```diff
@@ -301,10 +312,12 @@ def get_conn(client_type):
...
-    else:
-        username = config.get_cloud_config_value(
-            "username", get_configured_provider(), __opts__, search_global=False
-        )
+
+    username = config.get_cloud_config_value(
+        "username", get_configured_provider(), __opts__, search_global=False
+    )
+
+    if username:
```
Comment (waynew) [inline salt/cloud/clouds/azurearm.py:320]: This one should be much more straightforward - there should be a test that returns none for username and enforces that neither username or password was passed to get_client. And then a test that ensures that if the username was returned that whatever username/password values were provided *were* passed to get_client.

**Review by Akm0d — DISMISSED** [review]

File: salt/cloud/clouds/azurearm.py (hunk 1 above)
Comment (waynew) [reply salt/cloud/clouds/azurearm.py:228]: @nicholasmhughes Excellent job! I think we do still need the other test to be able to merge this, but that should be much more straightforward? 😂

**Review by waynew — CHANGES_REQUESTED** [review]
This looks great! I think just adding a changelog entry, perhaps `49063.added` with `Added support for MSI-style authentication for azurearm.` would be reasonable?

Thanks!

**Comment by nicholasmhughes** [comment]
@waynew unfortunately, that message would be misleading. the main reason for the original PR has since been merged for a while... I basically just ported over some additional housekeeping that got left behind. after all that, the main thing that's been added was tests... :laughing:  I don't even know what to call this change...

**Comment by waynew** [comment]
Well, it *does* have the `else` case, so maybe it's just a fixed? 🤷

**Comment by nicholasmhughes** [comment]
:shrug: took a stab at it

**Review by waynew — APPROVED** [review]
Good enough for me! That's at least enough information for someone reading the changelog to be able to say, "Ah, yeah, I need to know more about that". Which is the point of the changelog 😂 

🚀
